airdrop/airdrop.gno

```diff
@@ -0,0 +1,215 @@
+package airdrop
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/hex"
+	"errors"
+	"std"
+	"time"
+
+	"gno.land/p/demo/grc/grc20"
+	i256 "gno.land/p/demo/int256"
+	"gno.land/p/demo/merkle"
+	u256 "gno.land/p/demo/uint256"
+)
+
+var (
+	ErrInvalidProof              = errors.New("INVALID_PROOF")
+	ErrInvalidClaimsCount        = errors.New("INVALID_CLAIMS_COUNT")
+	ErrFirstClaimIsNot128        = errors.New("FIRST_CLAIM_MUST_BE_MULT_128")
+	ErrRefundableTimestampNotSet = errors.New("REFUNDABLE_TIMESTAMP_NOT_SET")
+	ErrTooEarlyToRefund          = errors.New("TOO_EARLY_TO_REFUND")
+)
+
+type Claim struct {
+	ID      uint64
+	Claimee string
+	Amount  *u256.Uint // ? u256
+}
+
+func (c Claim) Bytes() []byte {
+	var buf bytes.Buffer
+	idBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(idBytes, c.ID)
+	buf.Write(idBytes)
+	buf.Write([]byte(c.Claimee))
+	amountStr := c.Amount.ToString()
+	buf.Write([]byte(amountStr))
+	return buf.Bytes()
+}
+
+type Config struct {
+	Root                [32]byte
+	RefundableTimestamp uint64 // TODO: timelock
+	RefundTo            std.Address
+}
+
+type Airdrop struct {
+	token         grc20.IGRC20
+	config        Config
+	claimedBitmap map[uint64]*u256.Uint
+	merkleTree    *merkle.Tree
+}
+
+type IAirdrop interface {
+	GetToken() string
+	GetConfig() Config
+	Claim(claim Claim, proof [][]byte) bool
+	Claim128(claims []Claim, remainingProof [][]byte) uint8
+	IsClaimed(claimID uint64) bool
+	Refund()
+}
+
+// NewAirdrop creates a new Airdrop instance
+func NewAirdrop(token grc20.IGRC20, config Config, claims []Claim) (*Airdrop, error) {
+	// create Merkle tree
+	//
+	// need to know the number of claims to create the hashables slice
+	hashables := make([]merkle.Hashable, len(claims))
+	for i, claim := range claims {
+		hashables[i] = claim
+	}
+	tree := merkle.NewTree(hashables)
+
+	// verify that the computed root matches the provided root
+	// if hex.EncodeToString(tree.Root()) != hex.EncodeToString(config.Root[:]) {
+	// 	return nil, errors.New("computed root does not match provided root")
+	// }
+
+	return &Airdrop{
+		token:         token,
+		config:        config,
+		claimedBitmap: make(map[uint64]*u256.Uint),
+		merkleTree:    tree,
+	}, nil
+}
+
+// GetToken returns the token address
+// func (a *Airdrop) GetToken() std.Address {
+// 	// TODO: get its address
+// 	return a.token.GetAddress()
+// }
+
+// GetConfig returns the airdrop configuration
+func (a *Airdrop) GetConfig() Config {
+	return a.config
+}
+
+// IsClaimed checks if a claim has been made (claimed)
+func (a *Airdrop) IsClaimed(claimID uint64) bool {
+	word, index := a.claimIDToBitmapIndex(claimID)
+	bitmap, exists := a.claimedBitmap[word]
+	if !exists {
+		return false
+	}
+	mask := new(u256.Uint).Lsh(u256.NewUint(1), uint(index))
+	return new(u256.Uint).And(bitmap, mask).Cmp(u256.Zero()) != 0
```

Comment:
Since the `uint256` does not provide `Bits`-like function, I need to check the value by making an mask